sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -564,44 +587,90 @@ cdef class PairwiseDistancesArgKmin(PairwiseDistancesReduction):
+
+              - 'parallel_on_Y' dispatches chunks of Y uniformly on threads.
+              Each thread then iterates on all the chunks of X. This strategy is
+              embarrassingly parallel but uses intermediate datastructures
+              synchronisation.
+
+              - 'auto' relies on a simple heuristic to choose between
+              'parallel_on__X' and 'parallel_on_Y'.
+
+              - None (default) looks-up in scikit-learn configuration for
+              `pairwise_dist_parallel_strategy`, and use 'auto' if it is not set.
+
+        return_distance : boolean, default=False
+            Return distances between each X vector and its
+            argkmin if set to True.
+
         Returns
         -------
-        argkmin: PairwiseDistancesArgKmin
-            The suited PairwiseDistancesArgKmin implementation.
+            Indices of argkmin for each vector in X and its associated distances
+            if return_distance=True.
+
+        Notes
+        -----
+            This public classmethod is responsible for introspecting the arguments
+            values to dispatch to the private :meth:`PairwiseDistancesArgKmin._compute`
+            instance method of the most appropriate :class:`PairwiseDistancesArgKmin`
+            concrete implementation.
+
+            All temporarily allocated datastructures necessary for the concrete
+            implementation are therefore freed when this classmethod returns.
+
+            This allows entirely decoupling the interface entirely from the
+            implementation details whilst maintaining RAII.
         """
-        # This factory comes to handle specialisations.
-        if metric in ("euclidean", "sqeuclidean") and not issparse(X) and not issparse(Y):
+        # Note (jjerphan): Some design thoughts for future extensions.
+        # This factory comes to handle specialisations for the given arguments.
+        # For future work, this might can be an entrypoint to specialise operations
+        # for various back-end and/or hardware and/or datatypes, and/or fused
+        # {sparse, dense}-datasetspair etc.
+        if (
+            metric in ("euclidean", "sqeuclidean")
+                and not issparse(X)
+                and not issparse(Y)
+        ):
             use_squared_distances = metric == "sqeuclidean"
-            return FastEuclideanPairwiseDistancesArgKmin(
+            pda = FastEuclideanPairwiseDistancesArgKmin(
                 X=X, Y=Y, k=k,
                 use_squared_distances=use_squared_distances,
                 chunk_size=chunk_size,
+                strategy=strategy,
                 metric_kwargs=metric_kwargs,
             )
+        else: # Fall back on the default
+            pda = PairwiseDistancesArgKmin(
+                datasets_pair=DatasetsPair.get_for(X, Y, metric, metric_kwargs),
+                k=k,
+                chunk_size=chunk_size,
+                strategy=strategy,
+            )
```

Comment:
`PairwiseDistancesArgKmin` is already a net improvement in terms of performance right?

I would be happy to merge the original PR without `FastEuclideanPairwiseDistancesArgKmin` at first and have `FastEuclideanPairwiseDistancesArgKmin` be a follow up PR.

Reply:
> PairwiseDistancesArgKmin is already a net improvement in terms of performance right?

Kind of.

> I would be happy to merge the original PR without FastEuclideanPairwiseDistancesArgKmin at first and have FastEuclideanPairwiseDistancesArgKmin be a follow up PR.

We can do it as part of [the plan](https://github.com/scikit-learn/scikit-learn/pull/21462#issuecomment-993862788).